pilot/pkg/bootstrap/server.go

```diff
@@ -438,6 +512,95 @@ func (s *Server) initSecureGrpcServer(options *istiokeepalive.Options) error {
 	return nil
 }
 
+// initialize secureGRPCServer - using K8S DNS certs
+func (s *Server) initSecureGrpcServerDNS() error {
+	certDir := DNSCertDir
+
+	key := path.Join(certDir, constants.KeyFilename)
+	cert := path.Join(certDir, constants.CertChainFilename)
+
+	tlsCreds, err := credentials.NewServerTLSFromFile(cert, key)
+	// certs not ready yet.
+	if err != nil {
+		return err
+	}
+
+	// TODO: parse the file to determine expiration date. Restart listener before expiration
+	certificate, err := tls.LoadX509KeyPair(cert, key)
+	if err != nil {
+		return err
+	}
+
+	opts := s.grpcServerOptions(s.Args.KeepaliveOptions)
+	opts = append(opts, grpc.Creds(tlsCreds))
+	s.secureGRPCServerDNS = grpc.NewServer(opts...)
+	s.EnvoyXdsServer.Register(s.secureGRPCServerDNS)
+
+	s.secureHTTPServerDNS = &http.Server{
+		TLSConfig: &tls.Config{
+			Certificates: []tls.Certificate{certificate},
+			VerifyPeerCertificate: func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+				// For now accept any certs - pilot is not authenticating the caller, TLS used for
+				// privacy
+				return nil
+			},
+			NextProtos: []string{"h2", "http/1.1"},
+			ClientAuth: tls.NoClientCert, // auth will be based on JWT token signed by K8S
+		},
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if r.ProtoMajor == 2 && strings.HasPrefix(
+				r.Header.Get("Content-Type"), "application/grpc") {
+				s.secureGRPCServerDNS.ServeHTTP(w, r)
+			} else {
+				s.mux.ServeHTTP(w, r)
+			}
+		}),
+	}
+
+	// Default is 15012 - istio-agent relies on this as a default to distinguish what cert auth to expect
+	dnsGrpc := fmt.Sprintf(":%d", s.basePort+12)
+
+	// create secure grpc listener
+	secureGrpcListener, err := net.Listen("tcp", dnsGrpc)
+	if err != nil {
+		return err
+	}
+
+	s.addStartFunc(func(stop <-chan struct{}) error {
+		go func() {
+			if !s.waitForCacheSync(stop) {
+				return
+			}
+
+			log.Infof("starting K8S-signed grpc=%s", dnsGrpc)
+			go func() {
+				// This seems the only way to call setupHTTP2 - it may also be possible to set NextProto
+				// on a listener
+				err := s.secureHTTPServerDNS.ServeTLS(secureGrpcListener, "", "")
+				msg := fmt.Sprintf("Stoppped listening on %s", dnsGrpc)
+				select {
+				case <-stop:
+					log.Info(msg)
+					if s.Args.ForceStop {
+						s.secureGRPCServerDNS.Stop()
+					} else {
+						s.secureGRPCServerDNS.GracefulStop()
+					}
+					ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+					defer cancel()
+					_ = s.secureHTTPServerDNS.Shutdown(ctx)
+					s.secureGRPCServerDNS.Stop()
+				default:
```

Comment:
Won't this panic immediately?

Reply:
It's same code as the Spiffe listener... Doesn't seem to panic - but not sure why it's there either.

Reply:
In any case this runs after the listener stops - which should be at shutdown.